### PR TITLE
Feature integrating ros lift ctrl with wrp ros

### PR DESCRIPTION
**Pull request by Timothylee1**
Modified the src/peripheral/README.md, included action and service files, and lift_server_node .hpp and .cpp.
Removed the lift_client_node .hpp and .cpp.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -139,6 +142,14 @@ add_dependencies(wrp_sdk_ros ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORT
 add_executable(mobile_base_node src/mobile_base/mobile_base_node.cpp)
 target_link_libraries(mobile_base_node wrp_sdk_ros ${catkin_LIBRARIES})
 
+# add_executable(lift_client_node src/peripheral/lift_client_node.cpp)
+# target_link_libraries(lift_client_node ${catkin_LIBRARIES})
```
Comment (hanskw-weston) [inline CMakeLists.txt]: Please remove these comments

File: include/wrp_ros/peripheral/lift_server_node.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "wrp_sdk/peripheral/lift_controller.hpp"
+#include "wrp_sdk/interface/lift_interface.hpp"
+
+namespace westonrobot {
+class ActionServer {
+ public:
+  ActionServer();
+  ~ActionServer();
+
```
Comment (hanskw-weston) [inline include/wrp_ros/peripheral/lift_server_node.hpp]: This should not be called ActionServer. Should be called LiftControllerNode.

File: include/wrp_ros/peripheral/lift_server_node.hpp
Comment (hanskw-weston) [inline include/wrp_ros/peripheral/lift_server_node.hpp:1]: This file should be called lift_controller_node.hpp

File: include/wrp_ros/peripheral/lift_server_node.hpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  bool SrvCB(wrp_ros::LiftQuery::Request& req,
```
Comment (hanskw-weston) [inline include/wrp_ros/peripheral/lift_server_node.hpp]: Name should be more descriptive. 
```suggestion
  bool QueryCallback(wrp_ros::LiftQuery::Request& req,
```

File: src/peripheral/README.md [hunk 4]
```diff
@@ -55,4 +55,15 @@ ROS wrappers around wrp_sdk peripheral drivers.
...
+
+### lift_server_node
+
+| Service        | Type               | Description                              |
+| -------------- | ------------------ | ---------------------------------------- |
+| `lift_service` | wrp_ros::LiftQuery | Query for state of the provided lift id. |
+
+| Parameter     | Type   | Description                                                   |
+| ------------- | ------ | ------------------------------------------------------------- |
+| `device_path` | char*  | Path to microcontroller port.                                 |
```
Comment (hanskw-weston) [inline src/peripheral/README.md]:
```suggestion
| `device_path` | char*  | Path to lift controller port.                                 |
```

File: include/wrp_ros/peripheral/lift_server_node.hpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    } while (state.position != goal->position);
+
+    // Update result values
+    result.id = state.id;
+    result.position = state.position;
+    result.speed = state.speed;
+
+    // Update result for client and mark the goal as succeeded
+    Server_.setSucceeded(result);
+  }
```
Comment (hanskw-weston) [inline include/wrp_ros/peripheral/lift_server_node.hpp]: Implementation should not be in hpp file but in cpp file

File: src/peripheral/lift_server_node.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ActionServer::ActionServer()
+    : liftController_(),
+      Server_(
+          nh_, "lift_action",
+          [this](const wrp_ros::LiftControlGoalConstPtr& goal) {
+            this->ExecuteCB(goal);
+          },
+          false) {
+  const char* device_path = FindSerialPort();
+  liftController_.Connect(device_path, baud_rate);
```
Comment (hanskw-weston) [inline src/peripheral/lift_server_node.cpp]: You specified in readme that device_path and baud_rate are ROS parameters but are not actually exposing these as parameters.

**Review by hanskw-weston — CHANGES_REQUESTED** [review]

File: include/wrp_ros/peripheral/lift_controller_node.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                     wrp_ros::LiftQuery::Response& res);
+
+ private:
+  int baud_rate_ = 115200;
+  std::string device_path_ = "/dev/ttyUSB0";
+
+  ros::NodeHandle nh_;
+  ros::ServiceServer srvServer_;
+  actionlib::SimpleActionServer<wrp_ros::LiftControlAction> Server_;
+  westonrobot::LiftController liftController_;
```
Comment (hanskw4267) [inline include/wrp_ros/peripheral/lift_controller_node.hpp:38]: names should follow the style guide

```suggestion
  ros::ServiceServer query_server_;
  actionlib::SimpleActionServer<wrp_ros::LiftControlAction> lift_control_server_;
  westonrobot::LiftController lift_controller_;
```

Comment (hanskw-weston) [inline include/wrp_ros/peripheral/lift_controller_node.hpp]: names should follow the style guide

```suggestion
  ros::ServiceServer query_server_;
  actionlib::SimpleActionServer<wrp_ros::LiftControlAction> lift_control_server_;
  westonrobot::LiftController lift_controller_;
```

File: src/peripheral/lift_controller_node.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  liftController_.Connect(device_path_, baud_rate_);
+  if (!liftController_.IsOkay()) {
+    std::cerr << "Failed to connect to lift controller" << std::endl;
+    ros::shutdown();
+  }
+
+  srvServer_ = nh_.advertiseService("/lift_controller/service",
+                                    &LiftControllerNode::QueryCallback, this);
+  Server_.start(); 
+}
```
Comment (hanskw-weston) [inline src/peripheral/lift_controller_node.cpp]: Maybe move the Server_ below the connect. There is a case here that a action goal arrives before you connect to the lift successfully.

```suggestion
      {
  if (!ReadParameters()) {
    ROS_ERROR("Could not load parameters");
    ros::shutdown();
  }

  liftController_.Connect(device_path_, baud_rate_);
  if (!liftController_.IsOkay()) {
    std::cerr << "Failed to connect to lift controller" << std::endl;
    ros::shutdown();
  }

  srvServer_ = nh_.advertiseService("/lift_controller/service",
                                    &LiftControllerNode::QueryCallback, this);
                                   
  Server_ = actionlib::SimpleActionServer<wrp_ros::LiftControlAction>(
      nh_, "lift_controller_action",
      [this](const wrp_ros::LiftControlGoalConstPtr& goal) {
        this->LiftControllerCallback(goal);
      },
      false);
  Server_.start(); 
}
```

File: src/peripheral/README.md [hunk 9]
```diff
@@ -55,4 +55,15 @@ ROS wrappers around wrp_sdk peripheral drivers.
...
 | ------------- | ------ | ------------------------------------------------------------ |
-| `device_path` | string | Path to power regulator can port.<br />Default: "can0"<br /> |
\ No newline at end of file
+| `device_path` | string | Path to power regulator can port.<br />Default: "can0"<br /> |
+
+### lift_controller_node
+
+| Service                   | Type               | Description                              |
+| ------------------------- | ------------------ | ---------------------------------------- |
+| `/lift_controller/service`| wrp_ros::LiftQuery | Query for state of the provided lift id. |
```
Comment (hanskw-weston) [inline src/peripheral/README.md]: Please add the action into the readme and maybe rename the service and action topics to be more descriptive

maybe 

/lift_controller/state_query
/lift_controller/goal

**Review by hanskw-weston — CHANGES_REQUESTED** [review]

File: action/LiftControl.action [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#Goal 
+uint8 id
```
Comment (karthee-weston) [inline action/LiftControl.action:4]: Include comments to detail which id number corresponds to which lift.

**Review by karthee-weston — APPROVED** [review]
LGTM

**Review by karthee-weston — APPROVED** [review]
LGTM

**Review by karthee-weston — APPROVED** [review]
LGTM

**Pull request merged**